require(shiny)
require(reactable)
require(rclipboard)

shinyUI(
    fluidPage(
        ####################
        #### Custom CSS ####
        ####################

        tags$head(
            tags$style(HTML("
                        /* Smaller font for preformatted text */
                        pre, table.table {
                        font-size: smaller;
                        }

                        .shiny-notification {
                        height: 100px;
                        width: 400px;
                        position:fixed;
                        top: calc(100% - 50px);;
                        left: calc(100% - 400px);;
                        }

                        body {
                        min-height: 2000px;
                        }

                        .option-group {
                        border: 1px solid #ccc;
                        border-radius: 6px;
                        padding: 0px 5px;
                        margin: 5px -10px;
                        background-color: #f2f2f2;
                        }

                        .icon {
                        display:inline;
                        }

                        .option-header {
                        color: #79d;
                        text-transform: uppercase;
                        margin-bottom: 5px;
                        }
                        "))),

        rclipboardSetup(),

        # Application title
        titlePanel("TECHS - MS-DArT-seq visualization tool"),

        mainPanel(
            column(12,
                   tabsetPanel(

                       ### Comparison within environments ###

                       tabPanel("Functional annotation - comparison among clones within environment",
                                fluidRow(
                                    column(width = 2,
                                           div(class = "option-group",
                                               checkboxGroupInput("annot_clones",
                                                                  "Select the clones",
                                                                  choices = list("A1" = "A1",
                                                                                 "B2" = "B2",
                                                                                 "D4" = "D4",
                                                                                 "E5" = "E5",
                                                                                 "Q8" = "Q8"),
                                                                  selected = c("A1",
                                                                               "B2",
                                                                               "D4",
                                                                               "E5",
                                                                               "Q8")))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("annot_tissue",
                                                            "Selection of tissue",
                                                            choices = list("Leaves" = "leaf",
                                                                           "Xylem" = "wood"),
                                                            inline = F))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("upset_mod",
                                                            "Selection of UpSet behavior",
                                                            choices = list("Intersect" = "intersect",
                                                                           "Distinct" = "distinct"),
                                                            selected = "distinct"))),
                                    column(width = 2,
                                           conditionalPanel(
                                               condition = "input.annot_clones.length > 0",
                                               uiOutput('comps_to_annot_ui'))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("enrich_mod",
                                                            "For the enrichment analysis:",
                                                            choices = list("Use only genes that has at least one MSD-Site " = "msd_only",
                                                                           "Use all genes of the genome as background" = "all"),
                                                            inline = F)),
                                           div(class = "option-group",
                                               numericInput("enr_fdr_cutoff",
                                                            label = "Select significance level of the GO enrichment (FDR)",
                                                            value = 0.05)),
                                           div(class = "option-group",
                                               numericInput("enr_sim_cutoff",
                                                            label = "Select significance level to reduce the redundancy of the enriched GO terms (small value is more restrictive)",
                                                            value = 0.7)))),
                                fluidRow(
                                    titlePanel("UpSet plot showing the MSD-Methylated sites among clones in the Dry location (MG)"),
                                    column(width = 9,
                                           plotOutput("upset_g2_leaves", height = "600px"))),
                                fluidRow(
                                    titlePanel("UpSet plot showing the MSD-Methylated sites among clones in the Wet location (PR)"),
                                    column(width = 9,
                                           plotOutput("upset_g3_leaves", height = "600px"))),
                                fluidRow(
                                    titlePanel("Annotation of genes containing at least one of the MSD-Methylated sites of the selected subset of UpSet plot - Dry location (MG)"),
                                    column(width = 10,
                                           reactableOutput("annot_g2")),
                                    ## Set the clipboard
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("clip_annot_MG_filter",
                                                            "Select the data to copy to clipboard",
                                                            choices = list("Gene name" = "Gene name",
                                                                           "MSD-site name" = "MDS-sites",
                                                                           "GO ID" = "GO ID"),
                                                            inline = F))),
                                    column(width = 2,
                                           uiOutput("clip_annot_MG")),
                                    div(class = "option-group",
                                        column(width = 2,
                                               downloadButton('annot_g2_download',
                                                              HTML('Download a table with the annotation <br/> of methylated genes - Dry (MG)'),
                                                              style="background-color: #c2c2c2")),
                                        column(width = 2,
                                               actionButton("run_enrich_mg",
                                                            HTML("Run the enrichment analysis to the<br/>  methylated genes in MG (Dry)"),
                                                            style="background-color: #d6fffe")))),
                                fluidRow(
                                    titlePanel("GO terms enriched in the selected set of genes that have at least one of the MSD-methylated sites (MG)"),
                                    column(width = 10,
                                           reactableOutput("enrich_g2")),
                                    ## Set the clipboard
                                    column(width = 2,
                                           uiOutput("clip_enrich_MG")),
                                    div(class = "option-group",
                                        column(width = 2,
                                               downloadButton('enrich_g2_download',
                                                              HTML('Download a table with the enriched <br/> GO terms - Dry (MG)'),
                                                              style="background-color: #c2c2c2")))),

                                fluidRow(
                                    titlePanel("Annotation of genes containing at least one of the MSD-Methylated sites of the selected subset of UpSet plot - Wet location (PR)"),
                                    column(width = 10,
                                           reactableOutput("annot_g3")),
                                    ## Set the clipboard
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("clip_annot_PR_filter",
                                                            "Select the data to copy to clipboard",
                                                            choices = list("Gene name" = "Gene name",
                                                                           "MSD-site name" = "MDS-sites",
                                                                           "GO ID" = "GO ID"),
                                                            inline = F))),
                                    column(width = 2,
                                           uiOutput("clip_annot_PR")),
                                    div(class = "option-group",
                                        column(width = 2,
                                               downloadButton('annot_g3_download',
                                                              HTML('Download a table with the annotation<br/> of methylated genes - Wet (PR)'),
                                                              style="background-color: #c2c2c2")),
                                        column(width = 2,
                                               actionButton("run_enrich_pr",
                                                            HTML("Run the enrichment analysis to the<br/>  methylated genes in PR (Wet)"),
                                                            style="background-color: #d6fffe")))),
                                fluidRow(
                                    titlePanel("GO terms enriched in the selected set of genes that have at least one of the MSD-methylated sites (PR)"),
                                    column(width = 10,
                                           reactableOutput("enrich_g3")),
                                    ## Set the clipboard
                                    column(width = 2,
                                           uiOutput("clip_enrich_PR")),
                                    div(class = "option-group",
                                        column(width = 2,
                                               downloadButton('enrich_g3_download',
                                                              HTML('Download a table with the enriched<br/> GO terms - Wet (PR)'),
                                                              style="background-color: #c2c2c2"))))),

                       ### Comparison between environments ###

                       tabPanel("Functional annotation - Comparison between environments",
                                fluidRow(
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("venn_clones",
                                                            "Select the clones",
                                                            choices = list("A1" = "A1",
                                                                           "B2" = "B2",
                                                                           "D4" = "D4",
                                                                           "E5" = "E5",
                                                                           "Q8" = "Q8"),
                                                            selected = c("A1")))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("venn_tissue",
                                                            "Selection of tissue",
                                                            choices = list("Leaves" = "leaf",
                                                                           "Xylem" = "wood"),
                                                            inline = F))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("venn_enrich_mod",
                                                            "For the enrichment analysis:",
                                                            choices = list("Use only genes that has at least one MSD-Site " = "msd_only",
                                                                           "Use all genes of the genome as background" = "all"),
                                                            inline = F))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               numericInput("venn_enr_fdr_cutoff",
                                                            label = "Select significance level of the GO enrichment (FDR)",
                                                            value = 0.05))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               numericInput("venn_enr_sim_cutoff",
                                                            label = "Select significance level to reduce the redundancy of the enriched GO terms (small value is more restrictive)",
                                                            value = 0.7)))),
                                fluidRow(
                                    #titlePanel(h1("Comparison between environments")),
                                    column(width = 6,
                                           plotOutput("sites_venn")),
                                    column(width = 4,
                                           div(class = "option-group",
                                               radioButtons("venn_enrich_set",
                                                            "Select the set of the venn that should be annotated",
                                                            choices = list("Dry" = "Dry",
                                                                           "Wet" = "Wet",
                                                                           "Intersection" = "Intersection"),
                                                            selected = c("Intersection"))))),
                                fluidRow(
                                    column(width = 9,
                                           titlePanel(h3("Functional annotation of the methylated genes of the venn subset")),
                                           reactableOutput("venn_table")),
                                    ## Set the clipboard
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("clip_venn_filter",
                                                            "Select the data to copy to clipboard",
                                                            choices = list("Gene name" = "Gene name",
                                                                           "MSD-site name" = "MDS-sites",
                                                                           "GO ID" = "GO ID"),
                                                            inline = F))),
                                    column(width = 3,
                                           uiOutput("clip_venn"))),
                                fluidRow(
                                    column(width = 8,
                                           titlePanel(h3("GO enriched terms of the methylated genes of the venn subset")),
                                           reactableOutput("enrich_venn")),
                                    column(width = 3,
                                           actionButton("run_enrich_venn",
                                                        HTML("Run the enrichment analysis of the<br/> methylated genes of the venn subset!"),
                                                        style="background-color: #d6fffe")),
                                    column(width = 3,
                                           uiOutput("clip_venn_enrich")),
                                    div(class = "option-group",
                                        column(width = 3,
                                               downloadButton('enrich_venn_download',
                                                              HTML('Download a table with the annotation <br/> of enriched GO terms of the selected subset'),
                                                              style="background-color: #c2c2c2"))))),

                       tabPanel("LFMM: MSD-methylated sites vs phenotype",
                                fluidRow(
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("flmm_tissue",
                                                            "Selection of tissue",
                                                            choices = list("Leaves" = "leaves",
                                                                           "Xylem" = "xylem"),
                                                            inline = F))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("flmm_marks_selection",
                                                            "Select MSD-sites set:",
                                                            choices = list("All MSD-sites" = "all",
                                                                           "MSD-sites methylated in at least one of the clones" = "methy_true"),
                                                            selected = "methy_true"))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               numericInput("lfmm_pvalue",
                                                            label = "Select significance level",
                                                            value = 0.01)),
                                           div(class = "option-group",
                                               checkboxGroupInput("lfmm_phenot_cent",
                                                                  "Scaling and centering phenotypic values",
                                                                  choices = list("TRUE" = TRUE,
                                                                                 "FALSE" =  FALSE),
                                                                  selected = TRUE))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               checkboxGroupInput("traits",
                                                                  "Selects the traits to be tested",
                                                                  choices = list("Volume" = "Volume",
                                                                                 "Total_height" = "Total_height",
                                                                                 "DBH" = "DBH"),
                                                                  selected = c("Volume",
                                                                               "Total_height",
                                                                               "DBH"))))),
                                fluidRow(
                                    titlePanel("Scree plot to select the number of K"),
                                    column(width = 3,
                                           div(class = "option-group",
                                               numericInput("flmm_k",
                                                            label = "Select K (number of components) based on scree plot",
                                                            value = 4))),
                                    column(width = 5,
                                           plotOutput("scree_plot")
                                    )),
                                ## Manhattan plots
                                conditionalPanel(
                                    condition = "input.traits.includes('Volume')",
                                    fluidRow(
                                        titlePanel("LFMM results - Volume"),
                                        column(width = 5,
                                               plotOutput("manhattan_vol")),
                                        column(width = 7,
                                               plotOutput("bar_vol"))),
                                    fluidRow(
                                        column(width = 8,
                                               reactableOutput("table_vol")),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_clone_vol",
                                                                      "Filter the results by clone (Volume)",
                                                                      choices = list("A1" = "A1",
                                                                                     "B2" = "B2",
                                                                                     "D4" = "D4",
                                                                                     "E5" = "E5",
                                                                                     "Q8" = "Q8")))),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_class_vol",
                                                                      "Filter the results by class (Volume)",
                                                                      choices = list("Unmethylated in both localities" = "Unmethylated in both localities",
                                                                                     "Methylated in both localities" = "Methylated in both localities",
                                                                                     "Methylated only in Dry (MG)" = "Methylated only in Dry (MG)",
                                                                                     "Methylated only in Wet (PR)" = "Methylated only in Wet (PR)")
                                                   ))),
                                        column(width = 4,
                                               uiOutput("clip_volume")))),
                                conditionalPanel(
                                    condition = "input.traits.includes('Stem_dry_mass')",
                                    fluidRow(
                                        titlePanel("LFMM results - Stem dry mass"),
                                        column(width = 5,
                                               plotOutput("manhattan_mass")),
                                        column(width = 7,
                                               plotOutput("bar_mass"))),
                                    fluidRow(
                                        column(width = 8,
                                               reactableOutput("table_mass")),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_clone_mass",
                                                                      "Filter the results by clone (Dry mass)",
                                                                      choices = list("A1" = "A1",
                                                                                     "B2" = "B2",
                                                                                     "D4" = "D4",
                                                                                     "E5" = "E5",
                                                                                     "Q8" = "Q8")))),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_class_mass",
                                                                      "Filter the results by class (Dry mass)",
                                                                      choices = list("Unmethylated in both localities" = "Unmethylated in both localities",
                                                                                     "Methylated in both localities" = "Methylated in both localities",
                                                                                     "Methylated only in Dry (MG)" = "Methylated only in Dry (MG)",
                                                                                     "Methylated only in Wet (PR)" = "Methylated only in Wet (PR)")
                                                   ))),
                                        column(width = 4,
                                               uiOutput("clip_steam")))),
                                conditionalPanel(
                                    condition = "input.traits.includes('Total_height')",
                                    fluidRow(
                                        titlePanel("LFMM results - Total height"),
                                        column(width = 5,
                                               plotOutput("manhattan_height")),
                                        column(width = 7,
                                               plotOutput("bar_height"))),
                                    fluidRow(
                                        column(width = 8,
                                               reactableOutput("table_height")),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_clone_height",
                                                                      "Filter the results by clone (Total height)",
                                                                      choices = list("A1" = "A1",
                                                                                     "B2" = "B2",
                                                                                     "D4" = "D4",
                                                                                     "E5" = "E5",
                                                                                     "Q8" = "Q8")))),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_class_height",
                                                                      "Filter the results by class (Total height)",
                                                                      choices = list("Unmethylated in both localities" = "Unmethylated in both localities",
                                                                                     "Methylated in both localities" = "Methylated in both localities",
                                                                                     "Methylated only in Dry (MG)" = "Methylated only in Dry (MG)",
                                                                                     "Methylated only in Wet (PR)" = "Methylated only in Wet (PR)")
                                                   ))),
                                        column(width = 4,
                                               uiOutput("clip_height")))),
                                conditionalPanel(
                                    condition = "input.traits.includes('DBH')",
                                    fluidRow(
                                        titlePanel("LFMM results - Diameter at breast height"),
                                        column(width = 5,
                                               plotOutput("manhattan_dbh")),
                                        column(width = 7,
                                               plotOutput("bar_dbh"))),
                                    fluidRow(
                                        column(width = 8,
                                               reactableOutput("table_dbh")),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_clone_dbh",
                                                                      "Filter the results by clone (Height to live crown base)",
                                                                      choices = list("A1" = "A1",
                                                                                     "B2" = "B2",
                                                                                     "D4" = "D4",
                                                                                     "E5" = "E5",
                                                                                     "Q8" = "Q8")))),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_class_dbh",
                                                                      "Filter the results by class (Height to live crown base)",
                                                                      choices = list("Unmethylated in both localities" = "Unmethylated in both localities",
                                                                                     "Methylated in both localities" = "Methylated in both localities",
                                                                                     "Methylated only in Dry (MG)" = "Methylated only in Dry (MG)",
                                                                                     "Methylated only in Wet (PR)" = "Methylated only in Wet (PR)")
                                                   ))),
                                        column(width = 4,
                                               uiOutput("clip_dbh")))),
                                conditionalPanel(
                                    condition = "input.traits.includes('Height_bottom_live_crown')",
                                    fluidRow(
                                        titlePanel("LFMM results - Height to live crown base"),
                                        column(width = 5,
                                               plotOutput("manhattan_bottom")),
                                        column(width = 7,
                                               plotOutput("bar_bottom"))),
                                    fluidRow(
                                        column(width = 8,
                                               reactableOutput("table_bottom")),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_clone_bottom",
                                                                      "Filter the results by clone (Height to live crown base)",
                                                                      choices = list("A1" = "A1",
                                                                                     "B2" = "B2",
                                                                                     "D4" = "D4",
                                                                                     "E5" = "E5",
                                                                                     "Q8" = "Q8")))),
                                        column(width = 4,
                                               div(class = "option-group",
                                                   checkboxGroupInput("flmm_res_filter_by_class_bottom",
                                                                      "Filter the results by class (Height to live crown base)",
                                                                      choices = list("Unmethylated in both localities" = "Unmethylated in both localities",
                                                                                     "Methylated in both localities" = "Methylated in both localities",
                                                                                     "Methylated only in Dry (MG)" = "Methylated only in Dry (MG)",
                                                                                     "Methylated only in Wet (PR)" = "Methylated only in Wet (PR)")
                                                   ))),
                                        column(width = 4,
                                               uiOutput("clip_tblc"))))),

                       tabPanel("All genes of the genome annotation",
                                fluidRow(
                                    column(width = 2,
                                           div(class = "option-group",
                                               textInput("annot_filter_by_id",
                                                         "Insert the ID list! (Do not use quotes)",
                                                         ""))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("annot_filter_by_id_options",
                                                            "Select the type of ID used as input",
                                                            choices = list("No filter" = "all",
                                                                           "Gene name" = "Gene name",
                                                                           "MDS-site - Will show the annotation of this sites" = "MDS-Site",
                                                                           "GO ID" = "GO ID"),
                                                            selected = "all"))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               textInput("annot_filter_by_id_options_delimiter",
                                                         "Insert delimiter character",
                                                         ";"))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               radioButtons("venn_enrich_all_genes_mod",
                                                            "For the enrichment analysis:",
                                                            choices = list("Use only genes that has at least one MSD-Site " = "msd_only",
                                                                           "Use all genes of the genome as background" = "all"),
                                                            inline = F))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               numericInput("venn_enr_fdr_all_genes_cutoff",
                                                            label = "Select significance level of the GO enrichment (FDR)",
                                                            value = 0.05))),
                                    column(width = 2,
                                           div(class = "option-group",
                                               numericInput("venn_enr_sim_all_genes_cutoff",
                                                            label = "Select significance level to reduce the redundancy of the enriched GO terms (small value is more restrictive)",
                                                            value = 0.7)))),
                                fluidRow(
                                    column(width = 10,
                                           reactableOutput("all_genes_annotation")),
                                    column(width = 2,
                                           actionButton("run_enrich_all_genes",
                                                        HTML("Run the enrichment analysis to the<br/>  selected  genes (Dry)"),
                                                        style="background-color: #d6fffe"))),
                                fluidRow(
                                    titlePanel("Enriched GO terms of the selected genes"),
                                    column(width = 12,
                                           reactableOutput("enrich_venn_all_genes")))),
                       tabPanel("Methylation patterns of all MSD-sites",
                                fluidRow(
                                    column(width = 3,
                                           div(class = "option-group",
                                               textInput("mds_filter_by_id",
                                                         "Insert the list of MSD-sites to filter! (Do not use quotes",
                                                         ""))),
                                    column(width = 3,
                                           div(class = "option-group",
                                               textInput("mds_filter_by_id_options_delimiter",
                                                         "Insert delimiter character",
                                                         ";"))),
                                    column(width = 3,
                                           div(class = "option-group",
                                               radioButtons("heat_map_tissue",
                                                            "Selection of tissue",
                                                            choices = list("Leaves" = "leaf",
                                                                           "Xylem" = "wood"),
                                                            inline = F)))),
                                conditionalPanel(
                                    condition = "input.heat_map_tissue == 'leaf'",
                                    fluidRow(
                                        titlePanel("MSD-sites tested in leaves tissue"),
                                        column(width = 12,
                                               reactableOutput("all_marks_leaves"))),
                                    fluidRow(
                                        titlePanel("MSD-sites tested in leaves tissue - 'Heatmap'"),
                                        column(width = 5,
                                               plotOutput("selec_k_leaves_p1")),
                                        column(width = 5,
                                               plotOutput("selec_k_leaves_p2")),
                                        column(width = 3,
                                               div(class = "option-group",
                                                   numericInput("n_of_clusters",
                                                                "Insert the number of clusters based on the above plots",
                                                                ""))),
                                        column(width = 2,
                                               actionButton("run_clustered_heatmap_leaves",
                                                            HTML("Generate the clustered heatmap"),
                                                            style="background-color: #d6fffe"))),
                                    fluidRow(
                                        column(width = 12,
                                               plotOutput("clustered_heatmap_leaves", height = "1600px")),
                                        column(width = 3,
                                               div(class = "option-group",
                                                   radioButtons("filter_cluster_heatmap",
                                                                "Filter clusters to visualize?",
                                                                choices = list("Yes" = "yes",
                                                                               "No" = "no"),
                                                                selected = "no"),
                                                   textInput("clusters_to_keep",
                                                             "Insert the number of the clusters to keep (separated by comma if more than one)",
                                                             ""))))),
                                conditionalPanel(
                                    condition = "input.heat_map_tissue == 'wood'",
                                    fluidRow(
                                        titlePanel("MSD-sites tested in leaves tissue"),
                                        column(width = 12,
                                               reactableOutput("all_marks_leaves_x"))),
                                    fluidRow(
                                        titlePanel("MSD-sites tested in leaves tissue - 'Heatmap'"),
                                        column(width = 5,
                                               plotOutput("selec_k_leaves_p1_x")),
                                        column(width = 5,
                                               plotOutput("selec_k_leaves_p2_x")),
                                        column(width = 3,
                                               div(class = "option-group",
                                                   numericInput("n_of_clusters_x",
                                                                "Insert the number of clusters based on the above plots",
                                                                ""))),
                                        column(width = 2,
                                               actionButton("run_clustered_heatmap_leaves_x",
                                                            HTML("Generate the clusterd heatmap"),
                                                            style="background-color: #d6fffe"))),
                                    fluidRow(
                                        column(width = 12,
                                               plotOutput("clustered_heatmap_leaves_x", height = "1600px")),
                                        column(width = 3,
                                               div(class = "option-group",
                                                   radioButtons("filter_cluster_heatmap_x",
                                                                "Filter clusters to visualize?",
                                                                choices = list("Yes" = "yes",
                                                                               "No" = "no"),
                                                                selected = "no"),
                                                   textInput("clusters_to_keep_x",
                                                             "Insert the number of the clusters to keep (separated by comma if more than one)",
                                                             "")))))))),
            width = 12)
    )
)
